site/src/contexts/ProxyContext.test.tsx

```diff
@@ -0,0 +1,387 @@
+
+// fakeLatency is a helper function to make a Latency report from just a number.
+const fakeLatency = (ms: number): ProxyLatency.ProxyLatencyReport => {
+  return {
+    latencyMS: ms,
+    accurate: true,
+    at: new Date(),
+  }
+}
+
+describe("ProxyContextGetURLs", () => {
+  it.each([
+    ["empty", [], {}, undefined, "", ""],
+    // Primary has no path app URL. Uses relative links
+    [
+      "primary",
+      [MockPrimaryWorkspaceProxy],
+      {},
+      MockPrimaryWorkspaceProxy,
+      "",
+      MockPrimaryWorkspaceProxy.wildcard_hostname,
+    ],
+    [
+      "regions selected",
+      MockWorkspaceProxies,
+      {},
+      MockHealthyWildWorkspaceProxy,
+      MockHealthyWildWorkspaceProxy.path_app_url,
+      MockHealthyWildWorkspaceProxy.wildcard_hostname,
+    ],
+    // Primary is the default if none selected
+    [
+      "no selected",
+      [MockPrimaryWorkspaceProxy],
+      {},
+      undefined,
+      "",
+      MockPrimaryWorkspaceProxy.wildcard_hostname,
+    ],
+    [
+      "regions no select primary default",
+      MockWorkspaceProxies,
+      {},
+      undefined,
+      "",
+      MockPrimaryWorkspaceProxy.wildcard_hostname,
+    ],
+    // Primary is the default if the selected is unhealthy
+    [
+      "unhealthy selection",
+      MockWorkspaceProxies,
+      {},
+      MockUnhealthyWildWorkspaceProxy,
+      "",
+      MockPrimaryWorkspaceProxy.wildcard_hostname,
+    ],
+    // This should never happen, when there is no primary
+    ["no primary", [MockHealthyWildWorkspaceProxy], {}, undefined, "", ""],
+    // Latency behavior
+    [
+      "best latency",
+      MockWorkspaceProxies,
+      {
+        [MockPrimaryWorkspaceProxy.id]: fakeLatency(100),
+        [MockHealthyWildWorkspaceProxy.id]: fakeLatency(50),
+        // This should be ignored because it's unhealthy
+        [MockUnhealthyWildWorkspaceProxy.id]: fakeLatency(25),
+        // This should be ignored because it is not in the list.
+        ["not a proxy"]: fakeLatency(10),
+      },
+      undefined,
+      MockHealthyWildWorkspaceProxy.path_app_url,
+      MockHealthyWildWorkspaceProxy.wildcard_hostname,
+    ],
+  ])(
+    `%p`,
+    (
+      _,
+      regions,
+      latencies,
+      selected,
+      preferredPathAppURL,
+      preferredWildcardHostname,
+    ) => {
+      const preferred = getPreferredProxy(regions, selected, latencies)
+      expect(preferred.preferredPathAppURL).toBe(preferredPathAppURL)
+      expect(preferred.preferredWildcardHostname).toBe(
+        preferredWildcardHostname,
+      )
+    },
+  )
+})
+
+const TestingComponent = () => {
+  return renderWithAuth(
+    <ProxyProvider>
+      <TestingScreen />
+    </ProxyProvider>,
+    {
+      route: `/proxies`,
+      path: "/proxies",
+    },
+  )
+}
+
+// TestingScreen just mounts some components that we can check in the unit test.
+const TestingScreen = () => {
+  const { proxy, userProxy, isFetched, isLoading, clearProxy, setProxy } =
+    useProxy()
```

Comment:
You probably don't need this to test a hook. You can use the `renderHook` from `@testing-library/react`. I think it can make the test way easier to reason about. More about its usage: https://kentcdodds.com/blog/how-to-test-custom-react-hooks

Reply:
This looks like exactly what I was looking for.

In the example they have this:

```typescript
const {result} = renderHook(() => useUndo('one'))
```

The only caveat I have is I also need the `useDashboard` provider present until this leaves experimental. So I'd need to so `useProxy` with the Dashboard context available :thinking: